actions/recovery.go

```diff
@@ -26,8 +26,19 @@ import (
 	"github.com/pkg/errors"
 
 	"github.com/google/fscrypt/crypto"
+	"github.com/google/fscrypt/metadata"
 )
 
+// modifiedContextWithSource returns a copy of ctx with the protector source
+// replaced by source.
+func modifiedContextWithSource(ctx *Context, source metadata.SourceType) *Context {
+	modifiedConfig := *ctx.Config
+	modifiedConfig.Source = source
+	modifiedCtx := *ctx
+	modifiedCtx.Config = &modifiedConfig
+	return &modifiedCtx
```

Comment:
This isn't great, but is probably unavoidable without restructuring the underlying library to just not use `Context`s.

I'm wish Go supported Cloning a struct, as this code becomes hard to maintain and easy to break.